lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  /// exit block. This will not have any effect if there is a single exit
+  /// block already.
+  ///   \param exitIndices   Map from exit blocks to the assignedindex.
+  ///   \param exitIndexArg  SILValue that holds the exitIndex set in the loop.
+  SILBasicBlock *createNewExitBlockWithDemux(
+      const llvm::DenseMap<SILBasicBlock *, intmax_t> &exitIndices,
+      SILValue exitIndexArg);
+
+  /// Replace `termInst` with a new TermInst such that the edge to the target
+  /// consists of the original arguments and the given `newArgs` at the end.
+  /// The old termInst is removed and this returns the newly constructed one.
+  TermInst *appendArguments(TermInst *termInst, SILBasicBlock *target,
+                            ArrayRef<SILValue> newArgs);
+
+  /// Return an undef value of the given type.
+  SILValue getUndef(SILType type) const {
+    return SILUndef::get(type, currentFn->getModule());
+  }
+
+  // Configuration for graph construction.
+  GraphGlobalConfiguration *graphConfiguration;
+  DominanceInfo* DI;
+  SILLoopInfo* LI;
+  SILLoop *loop;
+  SILBasicBlock *header;
+  SILBasicBlock *preheader;
+  SILBasicBlock *latch;
+  SILFunction *currentFn;
+  /// exit blocks before the loop is transformed.
+  SmallVector<SILBasicBlock*, 8> exitBlocks;
+  /// The list of edges that need to rewired to the newHeader or
+  // the new latchBlock as appropriate.
+  SmallVector<std::pair<const SILBasicBlock *, const SILBasicBlock *>, 8>
+      edgesToFix;
+  /// Identify the set of values that escape the loop.
+  llvm::SmallPtrSet<SILValue, 8> escapingValues;
+};
+
+void SingleExitLoopTransformer::initialize() {
+  // Remember some information from the loop before it gets transformed.
+
+  // exit Blocks
+  loop->getExitBlocks(exitBlocks);
+  // All the exiting edges need to be rewired.
+  loop->getExitEdges(edgesToFix);
+  // All the edges to the header need to be rewired.
+  for (SILBasicBlock *headerPred : header->getPredecessorBlocks()) {
+    edgesToFix.emplace_back(headerPred, header);
+  }
+
+  for (const SILBasicBlock *bb : loop->getBlocks()) {
+    // Save the values that are escaping this loop in escapingValues set.
+    auto saveEscaping = [this](SILValue value) {
+      for (const auto *use : value->getUses()) {
+        const SILInstruction *useInst = use->getUser();
+        if (!this->loop->contains(useInst->getParent())) {
+          escapingValues.insert(value);
+          break;
+        }
+      }
+    };
+    if (bb != header) {
+      llvm::for_each(bb->getArguments(), saveEscaping);
+    }
+    for (const SILInstruction &inst : *bb) {
+      llvm::for_each(inst.getResults(), saveEscaping);
+    }
+  }
+}
+
+/// Appends the given arguments to the given edge. Deletes the old TermInst
+/// and returns a new TermInst with the appropriate number of args.
+TermInst *SingleExitLoopTransformer::appendArguments(
+    TermInst *termInst, SILBasicBlock *target, ArrayRef<SILValue> newArgs) {
+  SILBuilder builder(termInst);
```

Comment:
Since the entire transformation is local to a function, it would be better to define a single builder in the Transformer class.

Reply:
Ah, good idea. Will do in a subsequent PR as a cleanup.